diff --git a/src/airconv/airconv_public.h b/src/airconv/airconv_public.h
index cc1be7b..c7d7a99 100644
--- a/src/airconv/airconv_public.h
+++ b/src/airconv/airconv_public.h
@@ -5,7 +5,7 @@
 #ifndef __AIRCONV_H
 #define __AIRCONV_H
 
-#define AIRCONV_VERSION 32
+#define AIRCONV_VERSION 33
 
 #ifdef __cplusplus
 #include <string>
diff --git a/src/airconv/nt/dxbc_converter_base.cpp b/src/airconv/nt/dxbc_converter_base.cpp
index 7b42744..9ed4616 100644
--- a/src/airconv/nt/dxbc_converter_base.cpp
+++ b/src/airconv/nt/dxbc_converter_base.cpp
@@ -1660,6 +1660,9 @@ Converter::operator()(const InstSampleLOD &sample) {
   auto [Value, Residency] =
       air.CreateSample(Tex->Texture, Tex->Handle, SamplerHandle, Coord, ArrayIndex, sample.offsets, sample_level{LOD});
 
+  if (ctx.shader_type == microsoft::D3D10_SB_VERTEX_SHADER)
+    Value = ir.CreateFreeze(Value);
+
   StoreOperand(sample.dst, MaskSwizzle(Value, GetMask(sample.dst), Tex->Swizzle));
 }
 
diff --git a/src/d3d11/d3d11_context_impl.cpp b/src/d3d11/d3d11_context_impl.cpp
index 5b00f63..646a4e3 100644
--- a/src/d3d11/d3d11_context_impl.cpp
+++ b/src/d3d11/d3d11_context_impl.cpp
@@ -1225,7 +1225,8 @@ public:
         // Also MSDN: A resource cannot be used as a destination if: the resource is created with immutable or
         // dynamic usage.
         // So it's legal?
-        UNIMPLEMENTED("update buffer: staging");
+        memcpy(static_cast<uint8_t *>(staging->mappedImmediateMemory()) + copy_offset,
+               pSrcData, copy_len);
       } else if (auto bindable = GetResourceCommon(pDstResource)) {
         auto [staging_buffer, offset] = AllocateStagingBuffer(copy_len, 16);
         staging_buffer.updateContents(offset, pSrcData, copy_len);
